### PR TITLE
feat: (observability) add spans for BatchTransaction and Table

### PR DESCRIPTION
**Pull request by odeke-em**
This change is part of a series of changes to add
OpenTelemetry traces, focused on BatchTransaction and Table.

While here, made the tests for sessionPool spans much more precise to avoid flakes.

Updates #2079
Built from PR #2087
Updates #2114

### CODE REVIEW

File: src/table.ts [hunk 1]
```diff
@@ -1072,29 +1073,34 @@ class Table {
...
+            span.end();
+            callback(err);
+            return;
+          }
 
-        transaction![method](this.name, rows as Key[]);
-        transaction!.commit(options, callback);
-      }
-    );
+          span.end();
```
Comment (surbhigarg92) [inline src/table.ts:1098]: Why are we doing `span.end` before commit call ?

**Comment by surbhigarg92** [comment]
@odeke-em I don't see any tests for these spans.

**Comment by odeke-em** [comment]
Hey @surbhigarg92 I hadn't uploaded some of them as I created this PR early morning at around 3AM before bed and  in interludes. I've added comprehensive tests, please take a look.

File: src/transaction.ts [hunk 2]
```diff
@@ -285,6 +286,7 @@ export class Snapshot extends EventEmitter {
   queryOptions?: IQueryOptions;
   resourceHeader_: {[k: string]: string};
   requestOptions?: Pick<IRequestOptions, 'transactionTag'>;
+  observabilityOptions?: ObservabilityOptions;
```
Comment (surbhigarg92) [inline src/transaction.ts:289]: `observabilityOptions` should be injected by customers from [SpannerOptions](https://github.com/googleapis/nodejs-spanner/blob/main/src/index.ts#L134) which should set this property. Do you plan to do that in separate PR ? 
I would recommend to add that in the same PR, so that we can test batch-transaction and mutation flow end to end

File: test/batch-transaction.ts [hunk 3]
```diff
@@ -485,4 +491,147 @@ describe('BatchTransaction', () => {
...
+      batchTransaction.session = SESSION as {} as Session;
+      batchTransaction.id = ID;
+      batchTransaction.observabilityOptions = {tracerProvider: provider};
+
+      REQUEST.callsFake((_, callback) => callback(null, RESPONSE));
+    });
+
+    const GAX_OPTS = {};
+
+    const fakeDirectedReadOptionsForRequest = {
```
Comment (surbhigarg92) [inline test/batch-transaction.ts:531]: This is optional param , so better to remove it to keep the test short. 
Same goes for dataBoostEnabled

File: test/batch-transaction.ts [hunk 4]
```diff
@@ -485,4 +491,147 @@ describe('BatchTransaction', () => {
...
+          // Sort the spans by duration.
+          spans.sort((spanA, spanB) => {
+            spanA.duration < spanB.duration;
+          });
+
+          const gotSpanNames: string[] = [];
+          spans.forEach(span => {
+            gotSpanNames.push(span.name);
+          });
+          const wantSpanNames = [
```
Comment (surbhigarg92) [inline test/batch-transaction.ts:581]: nit: rename to `expectedSpans`

File: test/batch-transaction.ts [hunk 5]
```diff
@@ -485,4 +491,147 @@ describe('BatchTransaction', () => {
...
+            true,
+            'at least 2 spans expected'
+          );
+
+          // Sort the spans by duration.
+          spans.sort((spanA, spanB) => {
+            spanA.duration < spanB.duration;
+          });
+
+          const gotSpanNames: string[] = [];
```
Comment (surbhigarg92) [inline test/batch-transaction.ts:577]: nit: rename to `actualSpans` . Here and elsewhere.

File: test/batch-transaction.ts [hunk 6]
```diff
@@ -485,4 +491,147 @@ describe('BatchTransaction', () => {
...
+
+          const spans = traceExporter.getFinishedSpans();
+          assert.strictEqual(
+            spans.length >= 2,
+            true,
+            'at least 2 spans expected'
+          );
+
+          // Sort the spans by duration.
+          spans.sort((spanA, spanB) => {
```
Comment (surbhigarg92) [inline test/batch-transaction.ts:573]: Why do we need to sort ?

File: test/batch-transaction.ts [hunk 7]
```diff
@@ -485,4 +491,147 @@ describe('BatchTransaction', () => {
...
+        (err, part, resp) => {
+          assert.ifError(err);
+
+          traceExporter.forceFlush();
+
+          const spans = traceExporter.getFinishedSpans();
+          assert.strictEqual(
+            spans.length >= 2,
+            true,
+            'at least 2 spans expected'
```
Comment (surbhigarg92) [inline test/batch-transaction.ts:569]: We should do strictEqual check. We expect exactly 2 spans not at least 2.

File: test/batch-transaction.ts [hunk 8]
```diff
@@ -485,4 +491,147 @@ describe('BatchTransaction', () => {
...
+    it('createQueryPartitions', done => {
+      const REQUEST = sandbox.stub();
+      const response = {};
+      REQUEST.callsFake((_, callback) => callback(null, response));
+
+      const res = batchTransaction.createQueryPartitions(
+        QUERY,
+        (err, part, resp) => {
+          assert.ifError(err);
+
```
Comment (surbhigarg92) [inline test/batch-transaction.ts:562]: nit: remove empty lines

File: test/batch-transaction.ts [hunk 9]
```diff
@@ -485,4 +491,147 @@ describe('BatchTransaction', () => {
...
+          spans.sort((spanA, spanB) => {
+            spanA.duration < spanB.duration;
+          });
+
+          const gotSpanNames: string[] = [];
+          spans.forEach(span => {
+            gotSpanNames.push(span.name);
+          });
+          const wantSpanNames = [
+            'CloudSpanner.BatchTransaction.createPartitions_',
```
Comment (surbhigarg92) [inline test/batch-transaction.ts:582]: Is it possible to test that `createPartitions_` is a child span of `createQueryPartitions`

File: test/session-pool.ts [hunk 10]
```diff
@@ -1412,25 +1412,13 @@ describe('SessionPool', () => {
...
 
-        const spans = exporter.getFinishedSpans();
+        const events = span.events;
+        assert.strictEqual(!events, false, 'Events must be set');
         assert.strictEqual(
-          spans.length,
-          1,
-          'Exactly 1 span should have been exported'
-        );
-        assert.strictEqual(
```
Comment (surbhigarg92) [inline test/session-pool.ts:1421]: Why did we remove this check from here ?

File: test/batch-transaction.ts (hunk 7 above)
Comment (odeke-em) [reply test/batch-transaction.ts:569]: We need at least 2 because as we add in spans to other methods, the test shall need modification but sure I can modify it.

File: test/batch-transaction.ts (hunk 6 above)
Comment (odeke-em) [reply test/batch-transaction.ts:573]: Spans aren't guaranteed to be returned in the exact order of their exporting.

File: test/batch-transaction.ts (hunk 9 above)
Comment (odeke-em) [reply test/batch-transaction.ts]: Absolutely! Good point.

File: test/batch-transaction.ts [hunk 11]
```diff
@@ -485,4 +491,147 @@ describe('BatchTransaction', () => {
       assert.strictEqual(identifier.timestamp, TIMESTAMP);
     });
   });
+
+  describe('observability traces', () => {
```
[the remark this answers is not in the document]
Comment (odeke-em) [reply test/batch-transaction.ts]: Done, added for both batch-transaction.ts and table.ts

File: test/batch-transaction.ts (hunk 3 above)
Comment (odeke-em) [reply test/batch-transaction.ts]: Done, thanks!

File: test/batch-transaction.ts (hunk 5 above)
Comment (odeke-em) [reply test/batch-transaction.ts]: Done, thanks!

File: test/batch-transaction.ts (hunk 4 above)
Comment (odeke-em) [reply test/batch-transaction.ts]: Done, thanks!

File: test/batch-transaction.ts (hunk 7 above)
Comment (odeke-em) [reply test/batch-transaction.ts]: Done, thanks!

File: test/batch-transaction.ts (hunk 9 above)
Comment (odeke-em) [reply test/batch-transaction.ts]: Done, thanks!

File: test/batch-transaction.ts (hunk 8 above)
Comment (odeke-em) [reply test/batch-transaction.ts]: Done.

File: src/transaction.ts (hunk 2 above)
Comment (odeke-em) [reply src/transaction.ts:289]: I shall have it plumbed in progressively as we get there. The reason being that this PR is bound to balloon if I add it straight from the top inside `new Spanner({observabilityOptions: opts})`. Just the way for each part I have been adding it in and testing.

File: test/session-pool.ts (hunk 10 above)
Comment (odeke-em) [reply test/session-pool.ts:1421]: In the testing environment, when ran isolated, the checks pass but when ran as a whole this test flakes here and there, a potential problem with OpenTelemetry. Besides, the point of the code was to ensure that the parent span actually has the required annotations which we do have and now it is very well cut out just for that.

File: observability-test/batch-transaction.ts [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+    }).BatchTransaction;
+  });
+
+  beforeEach(() => {
+    batchTransaction = new BatchTransaction(SESSION as {} as Session);
+  });
+
+  afterEach(() => sandbox.restore());
+
+  describe('observability traces', () => {
```
Comment (surbhigarg92) [inline observability-test/batch-transaction.ts:125]: nit rename to `BatchTransaction traces`

File: test/session-pool.ts [hunk 13]
```diff
@@ -1367,34 +1352,12 @@ describe('SessionPool', () => {
   });
 
   describe('trace annotations on active span', () => {
```
Comment (surbhigarg92) [inline test/session-pool.ts:1354]: maybe for separate PR. But It will be good to extract these tests in `observability-test` folder

**Review by surbhigarg92 — APPROVED** [review]

File: observability-test/batch-transaction.ts (hunk 12 above)
Comment (odeke-em) [reply observability-test/batch-transaction.ts]: Done, I instead removed the describe given that I moved the code from where I had written it nested inside test/batch-transaction.ts to here. This now what the log looks like
Screenshot 2024-09-19 at 7 36 26 PM
